### PR TITLE
Memory safety proofs for new aws_byte_buf functions

### PR DESCRIPTION
**Pull request by feliperodri**
*Description of changes:*

We now have a set of memory safety proofs for `aws_byte_buf_append_dynamic`, `aws_byte_buf_append_with_lookup` and `aws_byte_buf_reserve`.

We also added guards in the functions, in order to make sure they will always operate over valid `aws_byte_buf` structures.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by tautschnig** [comment]
This does not seem to be a clean branch as it includes a pre-existing commit.

File: .cbmc-batch/include/proof_helpers/make_common_data_structures.h [hunk 1]
```diff
@@ -37,6 +37,26 @@
 #define ASSUME_BOUNDED_ARRAY_LIST(list, max_initial_item_allocation, max_item_size)                                    \
     list = make_bounded_array_list((max_initial_item_allocation), (max_item_size))
 
+/*
+ * aws_byte_buf
```
Comment (tautschnig) [inline .cbmc-batch/include/proof_helpers/make_common_data_structures.h]: I believe this and the comment below could be a bit more verbose.

File: .cbmc-batch/jobs/aws_byte_buf_append_with_lookup/cbmc-batch.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+jobos: ubuntu16
+cbmcflags: "--bounds-check;--pointer-check;--div-by-zero-check;--signed-overflow-check;--unsigned-overflow-check;--pointer-overflow-check;--undefined-shift-check;--float-overflow-check;--nan-check;--unwinding-assertions;--function;aws_byte_buf_append_with_lookup_harness"
```
Comment (tautschnig) [inline .cbmc-batch/jobs/aws_byte_buf_append_with_lookup/cbmc-batch.yaml]: Missing unwindset

File: .cbmc-batch/jobs/aws_byte_buf_append_with_lookup/aws_byte_buf_append_with_lookup_harness.c [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/byte_buf.h>
+#include <proof_helpers/make_common_data_structures.h>
+
+#define MAX_BUF_SIZE 50
```
Comment (tautschnig) [inline .cbmc-batch/jobs/aws_byte_buf_append_with_lookup/aws_byte_buf_append_with_lookup_harness.c]: Why is this the correct magic number?

File: .cbmc-batch/source/make_common_data_structures.c [hunk 4]
```diff
@@ -21,6 +21,42 @@
...
 
+bool is_bounded_byte_buf(struct aws_byte_buf *buf, size_t max_size) {
+    return buf->capacity <= max_size;
+}
+
+bool is_valid_byte_buf(struct aws_byte_buf *buf) {
+    return (buf->len <= buf->capacity)
+    && (buf->allocator != NULL);
+    /* TODO: once we can check allocated size of buffer, also assert that
+     * buf->capacity <= __CPROVER_buffer_size(buf->buffer);
```
Comment (tautschnig) [inline .cbmc-batch/source/make_common_data_structures.c]: I believe you just need `__CPROVER_r_ok(buf->buffer, buf->capacity)`

File: .cbmc-batch/source/make_common_data_structures.c [hunk 5]
```diff
@@ -21,6 +21,42 @@
...
+}
+
+bool is_bounded_byte_cursor(struct aws_byte_cursor *cursor, size_t max_size) {
+    return cursor->len <= max_size;
+}
+
+bool is_valid_byte_cursor(struct aws_byte_cursor *cursor) {
+    return (cursor->ptr != NULL && cursor->len != 0);
+    /* TODO: once we can check allocated size of buffer, also assert that
+     * cursor->len <= __CPROVER_buffer_size(cursor->ptr);
```
Comment (tautschnig) [inline .cbmc-batch/source/make_common_data_structures.c]: As above: `__CPROVER_r_ok(cursor->ptr, cursor->len)`

File: .cbmc-batch/source/proof_allocators.c [hunk 6]
```diff
@@ -38,6 +38,12 @@ void *can_fail_malloc(size_t size) {
     return (nondet) ? NULL : malloc(size);
 }
 
+void *bounded_malloc(size_t size)  {
+    void* rval = malloc(size);
+    __CPROVER_assume(size < SIZE_MAX - (size_t)(rval));
```
Comment (tautschnig) [inline .cbmc-batch/source/proof_allocators.c]: Could you elaborate on `(size_t)(rval)` (or the subtraction thereof)?

File: .cbmc-batch/jobs/aws_byte_buf_append_with_lookup/aws_byte_buf_append_with_lookup_harness.c [hunk 7]
```diff
@@ -29,7 +27,8 @@ void aws_byte_buf_append_with_lookup_harness() {
     __CPROVER_assume(is_valid_byte_cursor(&from));
     ensure_byte_cursor_has_allocated_buffer_member(&from);
 
-    aws_byte_buf_append_with_lookup(&to, &from, aws_lookup_table_to_lower_get());
+    uint8_t* lookup_table[256];
```
Comment (tautschnig) [inline .cbmc-batch/jobs/aws_byte_buf_append_with_lookup/aws_byte_buf_append_with_lookup_harness.c]: Why is 256 the right number?

File: source/byte_buf.c [hunk 8]
```diff
@@ -514,7 +513,7 @@ int aws_byte_buf_reserve(struct aws_byte_buf *buffer, size_t requested_capacity)
     buffer->buffer = new_buffer;
     buffer->capacity = requested_capacity;
 
-    AWS_FATAL_ASSERT(is_valid_byte_buf(buffer));
+    assert(is_valid_byte_buf(buffer));
```
Comment (tautschnig) [inline source/byte_buf.c]: I'd squash this commit with the commit introducing the changes in the first place. Otherwise it makes for a messy git history.

File: .cbmc-batch/include/proof_helpers/make_common_data_structures.h [hunk 9]
```diff
@@ -37,6 +37,22 @@
...
+
+bool is_byte_buf_expected_alloc(struct aws_byte_buf *buf);
+
+void ensure_byte_buf_has_allocated_buffer_member(struct aws_byte_buf *buf);
+
+/*
+ * aws_byte_cursor
+ */
+bool is_bounded_byte_cursor(struct aws_byte_cursor *cursor, size_t max_size);
+
```
Comment (danielsn) [inline .cbmc-batch/include/proof_helpers/make_common_data_structures.h]: Add a comment

File: .cbmc-batch/include/proof_helpers/proof_allocators.h [hunk 10]
```diff
@@ -32,6 +32,8 @@ static void *can_fail_malloc_allocator(struct aws_allocator *allocator, size_t s
 
 void *can_fail_malloc(size_t size);
 
+void *bounded_malloc(size_t size);
```
Comment (danielsn) [inline .cbmc-batch/include/proof_helpers/proof_allocators.h]: Add a comment

File: cmake/AwsLibFuzzer.cmake [hunk 11]
```diff
@@ -21,7 +21,8 @@ set(FUZZ_TESTS_MAX_TIME 60 CACHE STRING "Max time to run each fuzz test")
 # Options:
 #  fuzz_files: The list of fuzz test files
 #  other_files: Other files to link into each fuzz test
```
Comment (danielsn) [inline cmake/AwsLibFuzzer.cmake]: This file should not be included in the PR.  Possibly needs a rebase.

File: tests/CMakeLists.txt [hunk 12]
```diff
@@ -294,4 +294,4 @@ if (NOT MSVC)
 endif()
 
 file(GLOB FUZZ_TESTS "fuzz/*.c")
-aws_add_fuzz_tests("${FUZZ_TESTS}" "")
+aws_add_fuzz_tests("${FUZZ_TESTS}" "" "")
```
Comment (danielsn) [inline tests/CMakeLists.txt]: This file should not be included in the PR.  Possibly needs a rebase.

**Comment by danielsn** [comment]
The overall proofs look good.  Just a few small comments.

File: .cbmc-batch/jobs/aws_byte_buf_reserve/aws_byte_buf_reserve_harness.c [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+#include <aws/common/byte_buf.h>
+#include <proof_helpers/make_common_data_structures.h>
+
+void aws_byte_buf_reserve_harness() {
+    struct aws_byte_buf buf;
+    __CPROVER_assume(is_valid_byte_buf(&buf));
+    ensure_byte_buf_has_allocated_buffer_member(&buf);
+
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_byte_buf_reserve/aws_byte_buf_reserve_harness.c]: Is it possible to add an extra assertion that if the reserve succeeds, the size went up, and if it fails, the bytes didn't change?
